asynciterator.ts

```diff
@@ -1966,3 +1966,53 @@ type SourceExpression<T> =
 
 type InternalSource<T> =
   AsyncIterator<T> & { _destination: AsyncIterator<any> };
+
+/**
+ * @param source An AsyncIterator
+ * @returns The AsyncIterator if it is not empty, otherwise undefined
+ */
+export async function maybeIterator<T>(source: AsyncIterator<T>): Promise<null | AsyncIterator<T>> {
+  // Avoid creating a new iterator where possible
+  if ((source instanceof ArrayIterator || source instanceof BufferedIterator) && (source as any)._buffer?.length > 0)
+    return source;
+
+  if (source instanceof IntegerIterator && (source as any)._step >= 0 ? (source as any)._next <= (source as any)._last : (source as any)._next >= (source as any)._last)
+    return source;
+
+
+  let item;
+  do {
+    if ((item = source.read()) !== null)
+      return source.prepend([item]);
```

Comment:
I would return `new AsyncIterator` whose `read` method has `if (item) { const rest = item; item = null; return item;} return iterator.read();`.

Reply:
Once https://github.com/RubenVerborgh/AsyncIterator/pull/65 is resolved/merged I imagine this doesn't produce much of an optimization, so not sure if it is worthwhile doing this.

Reply:
> local implementation for the quick one-element case.

In going to do this I realised that the logic `readable` events should be roughly identical to in https://github.com/RubenVerborgh/AsyncIterator/pull/59 so will wait until that is stable / merged and then we can use the same logic or refactor the necessary logic into a common abstract class.